src/cached_object_store/stats.rs

```diff
@@ -0,0 +1,45 @@
+use std::sync::Arc;
+use crate::stats::{Counter, Gauge, StatRegistry};
+
+macro_rules! oscache_stat_name {
+    ($suffix:expr) => {
+        crate::stat_name!("gc", $suffix)
+    };
+}
+
+pub const OBJECT_STORE_CACHE_PART_HITS: &'static str = oscache_stat_name!("part_hits");
+pub const OBJECT_STORE_CACHE_PART_ACCESS: &'static str = oscache_stat_name!("part_access");
+pub const OBJECT_STORE_CACHE_KEYS: &'static str = oscache_stat_name!("cache_keys");
+pub const OBJECT_STORE_CACHE_BYTES: &'static str = oscache_stat_name!("cache_bytes");
+pub const OBJECT_STORE_CACHE_EVICTED_KEYS: &'static str = oscache_stat_name!("evicted_keys");
+pub const OBJECT_STORE_CACHE_EVICTED_BYTES: &'static str = oscache_stat_name!("evicted_bytes");
+
+#[derive(Debug, Clone)]
+pub(crate) struct CachedObjectStoreStats {
+    pub(super) object_store_cache_part_hits: Arc<Counter>,
+    pub(super) object_store_cache_part_access: Arc<Counter>,
+    pub(super) object_store_cache_keys: Arc<Gauge<u64>>,
+    pub(super) object_store_cache_bytes: Arc<Gauge<u64>>,
+    pub(super) object_store_cache_evicted_keys: Arc<Counter>,
+    pub(super) object_store_cache_evicted_bytes: Arc<Counter>,
+}
+
+impl CachedObjectStoreStats {
+    pub(crate) fn new(registry: &StatRegistry) -> Self {
+        let stats = Self {
+            object_store_cache_part_hits: Arc::new(Counter::default()),
+            object_store_cache_part_access: Arc::new(Counter::default()),
+            object_store_cache_bytes: Arc::new(Gauge::default()),
+            object_store_cache_keys: Arc::new(Gauge::default()),
+            object_store_cache_evicted_bytes: Arc::new(Counter::default()),
+            object_store_cache_evicted_keys: Arc::new(Counter::default()),
+        };
+        registry.register(OBJECT_STORE_CACHE_PART_HITS, stats.object_store_cache_part_hits.clone());
```

Comment:
Feels weird to have the store stats register itself with the registry. I expected registry.register(store_stats). I guess this might be a rust thing, though. It doesn't seem like there's a nice way to have the register() method get the stat fields from the stats store unless return a map or something...

Reply:
Yeah it would have to return some collection. FWIW this is the pattern that the Kafka `Metrics` registry uses, so that's another data point.